plugin/meta_data_store.go

```diff
@@ -0,0 +1,53 @@
+package plugin
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+type metaDataStore struct {
+	dir           string
+	installTarget *installTarget
+}
+
+var errDisaleMetaDataStore = errors.New("MetaData disabled. could not detect owner/repo")
+
+func newMetaDataStore(pluginDir string, it *installTarget) (*metaDataStore, error) {
+	owner, repo, err := it.getOwnerAndRepo()
+	if err != nil {
+		return nil, errDisaleMetaDataStore
+	}
```

Comment:
For example `mkr plugin install mackerel-plugin-sample@v0.0.1`,

```go
it, _ := newInstallTargetFromString("mackerel-plugin-sample@v0.0.1")
// &plugin.installTarget{owner:"", repo:"", pluginName:"mackerel-plugin-sample", releaseTag:"v0.0.1", directURL:"", rawGithubURL:"", apiGithubURL:""}

it.getOwnerAndRepo()
// &plugin.installTarget{owner:"mackerelio", repo:"mackerel-plugin-sample", pluginName:"mackerel-plugin-sample", releaseTag:"v0.0.1", directURL:"", rawGithubURL:"", apiGithubURL:""}
```

*installTarget which created by `newInstallTargetFromString()` has not owner and repo in this case (`Plugin name only` and `Plugin name and release tag` in test cases).
After `it.getOwnerAndRepo()` called, owner and repo are filled by informations from plugin-registry.

So, I think `hasOwnerAndRepo` results is undetermined before `getOwnerAndRepo()` called.

Therefore `hasOwnerAndRepo()` becomes only a thin wrapper of `getOwnerAndRepo()` as below.
Is this need...?

```go
func (it *installTarget) hasOwnerAndRepo() (bool, error) {
    owner, repo, err := it.getOwnerAndRepo()
    return (owner != "" && repo != ""), err
}
```

Reply:
ah... I see.

I reconsider it.  The problem is
- if we call `getOwerAndRepo()` when installTarget have directURL
    - `mkr plugin install https://github.com/mackerelio/mackerel-plugin-sample/releases/download/v0.0.3/mackerel-plugin-sample_darwin_amd64.zip`
    - We can specify URL as install_target, but it is undocumented yet...  Sorry...
    - See also: https://github.com/mackerelio/mkr/pull/151
- then `getOwnerAndRepo()` access wrong URL unexpectedly
    - because of https://github.com/mackerelio/mkr/blob/master/plugin/install_target.go#L97..L101
    - e.g. https://raw.githubusercontent.com/mackerelio/plugin-registry/master/plugins/.json

So how do we fix `getOwnerAndRepo` like below?

```diff
diff --git a/plugin/install_target.go b/plugin/install_target.go
index 5c3a0ff..556cc1f 100644
--- a/plugin/install_target.go
+++ b/plugin/install_target.go
@@ -93,6 +93,11 @@ func (it *installTarget) getOwnerAndRepo() (string, string, error) {
 		return it.owner, it.repo, nil
 	}
 
+	// if directURL is specified, target doesn't have owner and repo
+	if it.directURL != "" {
+		return "", "", fmt.Errorf("owner and repo are not found because directURL is specified")
+	}
+
 	// Get owner and repo from plugin registry
 	defURL := fmt.Sprintf(
 		"%s/mackerelio/plugin-registry/master/plugins/%s.json",

```

Reply:
fixed in 3c56581baa84bd80bf8231c84298335709e104ad